api/document/models.py

```diff
@@ -19,3 +23,57 @@ class Meta:
         index_together = (
             unique_together,
         )
+
+    @staticmethod
+    def new_tree(tag: str, number: str, **attrs):
+        tree = DiGraph()
+        identifier = f"{tag}_{number}"
+        tree.add_node(identifier, model=DocNode(
+            identifier=identifier, tag=tag, number=number, depth=0,
+            **attrs
+        ))
+        return DocCursor(tree, identifier)
+
+
+class DocCursor():
+    """DocNodes don't keep track of their children/relationships within a
+    tree, so we will generally access them indirectly through a wrapping tree
+    (a DiGraph). This DocCursor points to a specific node within that tree."""
+    __slots__ = ('tree', 'identifier')
+
+    def __init__(self, tree: DiGraph, identifier: str):
+        self.tree = tree
+        self.identifier = identifier
+
+    @property
+    def model(self):
+        return self.tree.node[self.identifier]['model']
+
+    def __getitem__(self, child_suffix: str):
+        child_id = f"{self.identifier}__{child_suffix}"
+        if child_id not in self.tree:
+            raise KeyError(f"No {child_id} element")
+        return self.__class__(self.tree, child_id)
+
+    def children(self):
+        edges = [(sort_order, idx) for _, idx, sort_order
+                 in self.tree.out_edges(self.identifier, data='sort_order')]
+        for _, identifier in sorted(edges):
+            yield self.__class__(self.tree, identifier=identifier)
+
+    def next_number(self, tag):
+        child_tag_counts = Counter(c.model.tag for c in self.children())
+        return str(child_tag_counts[tag] + 1)
+
+    def add_child(self, tag: str, number: Optional[str] = None, **attrs):
```

Comment:
I'm slightly confused by `number` here; why is it a string rather than an integer? Perhaps the question is also “why is it called number?”, which would also apply to the name of the `DocNode` property.

Reply:
The idea is that we need to distinguish between multiple of the same tag within a parent, e.g. multiple sections within a subpart, multiple paragraphs within a section. The tricky bit is that, to retain semantic value, these might not always be integers. For example, we'd want `subpart_A`, `par_b__subpar_4__subsubpar_iv`, and similar. The term "number" comes from Akoma Ntoso, but we're not quite using it the way they are, so we can definitely rename. What do you think makes sense?

`marker` could make sense, but then it'd conflict with the text "(a)"/"Section 12"/etc., should we ever want to add that separately. Maybe `tag_idx`?

Reply:
What about `identifier`?